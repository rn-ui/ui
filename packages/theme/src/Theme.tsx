/* tslint:disable */
import deepMerge from './tools/deepMerge';
import mergeComponentAndThemes from './tools/mergeComponentAndThemes';
import { resolveIncludes } from './tools/resolveIncludes';
import normalizeTheme from './normalizer/normalizeTheme';
import defaultTheme from './resources/default';

interface Themes {
  default: object;
  light: object;
  dark: object;
}

// @ts-ignore
const themes: Themes = {};

// Privates, ideally those should be symbols
const THEME_STYLES = '@southem.theme';

const resolveStyle = (style, baseStyle) => normalizeTheme(resolveIncludes(style, baseStyle));

/**
 * Theme
 *
 * @example
 * const defaultTheme = {};
 * const defaultThemePro = {};
 *
 * Theme.registerTheme('default', [
 *   defaultTheme,
 *   defaultThemePro,
 * ]);
 */
class Theme {

  constructor(themeStyle) {
    // @ts-ignore
    // this[THEME_STYLES] = resolveStyle(themeStyle);
  }

  /**
   * registerTheme
   * @param {String} themeName
   * @param {Array|Object} styleSheets
   */
  static registerTheme(themeName, styleSheets: Array<Object>|Object): void {
    let mergedStyleSheets = styleSheets;

    if (Array.isArray(styleSheets)) {
      mergedStyleSheets = deepMerge({}, ...styleSheets);
    }

    // themes[themeName] = new Theme(themeName, mergedStyleSheets);
    // @ts-ignore
    themes[themeName] = resolveStyle(mergedStyleSheets);
  }

  /**
   * registerDefaultTheme
   * @param {Object} styleSheets
   */
  static registerDefaultTheme(styleSheets): void {
    Theme.registerTheme('default', styleSheets);
  }

  /**
   * getTheme
   * @param {String} themeName
   * @return {Object}
   */
  static getTheme(themeName): Object {
    // @ts-ignore
    if (themeName === 'default' && !themes.default) {
      Theme.registerDefaultTheme(defaultTheme);
    }

    return themes[themeName];
  }

  /**
   * getDefaultTheme
   * @return {Object}
   */
  static getDefaultTheme(): Object {
    return Theme.getTheme('default');
  }

  /**
   * isDefine
   * @param {String} themeName
   * @return {Boolean}
   */
  static isDefine(themeName): Boolean {
    return !!themes[themeName];
  }
}

export {
  Theme as default,
  themes,
};
