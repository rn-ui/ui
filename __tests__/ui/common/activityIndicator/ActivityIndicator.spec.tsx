import React from 'react';
import {
  StyleSheet,
  ActivityIndicator as RNActivityIndicator,
} from 'react-native';
import {
  render,
  RenderAPI,
  shallow,
} from 'react-native-testing-library';
import renderer, { ReactTestInstance } from 'react-test-renderer';
import Theme, { ThemeProvider } from '@southem/theme';
import {
  darkTheme,
  ActivityProps,
  ActivityIndicator,
} from '@southem/ui';

Theme.registerDefaultTheme(darkTheme);

const Mock = (props?: ActivityProps): React.ReactElement<{}> => {
  return (
    <ThemeProvider
      theme={'default'}>
      <ActivityIndicator {...props} />
    </ThemeProvider>
  );
};

const renderComponent = (props?: ActivityProps): RenderAPI => {
  return render(
    <Mock {...props} />,
  );
};

describe('@activityIndicator: matches snapshot', () => {
  describe('* appearance', () => {
    it('* default', () => {
      const component: RenderAPI = renderComponent({
        size: 'small',
      });

      const {output} = shallow(component.getByType(RNActivityIndicator));

      expect(output).toMatchSnapshot();
    });
  });
});

describe('@activityIndicator: component checks', () => {

  it('* color size styled properly', () => {
    const component: RenderAPI = renderComponent({
      size: 'small',
    });

    const activity: ReactTestInstance = component.getByType(RNActivityIndicator);
    const { color, size } = StyleSheet.flatten(activity.props);

    expect(size).toEqual('small');
    expect(color).toEqual('rgba(0, 0, 0, 0.5)');
  });
});
