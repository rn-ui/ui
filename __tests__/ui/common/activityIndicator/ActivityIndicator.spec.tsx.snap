// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@activityIndicator: matches snapshot * appearance * default 1`] = `
<ActivityIndicator
  animating={true}
  color="rgba(0, 0, 0, 0.5)"
  hidesWhenStopped={true}
  size="small"
  style={Object {}}
/>
`;
